Gf 2/feature

saykaren opened the pull request.

Commit:
removed link to favicon

Commit:
adding API

Commit:
adding a few items for the API but going to try on a different app first

Commit:
I believe I got a connection with the server on http://localhost:5000/user, will work on more details later

patrickrauls commented on src/Components/App/App.js:
```diff
@@ -9,16 +9,37 @@ import AboutMe from './AboutUser';
   const [friends, setFriends] = useState(userData[0].friends);
+  const [state, setState] = useState(null);
+
+  const componentDidMount = () =>{
```
consider using a useEffect for this https://reactjs.org/docs/hooks-effect.html

saykaren replied:
Thanks Patrick! I did get it going on my BomdigityAPI successfully so I am working on the GitFitter API as we speak to convert what I did sucessfully with BomdigityAPI onto here which is using the useEffects and functional components. So far at work today it is a bit slow so I should be able to get it done otherwise I have sometime this weekend slated for it. I will let you know if I bump into any issues

Commit:
added API and functional component to grab informmation

Commit:
API is working with all the other components but not when I want to pass it through to FriendsLookUp. I also want to move that to possibly the App.js as once I figure that out I would like to pass those names to the Exercise component as well

Commit:
updated data from API to come in as an array. Fixed data to adjust to that change. Updated FriendLookUp to show names not numbers. Put a FriendsName array as state in App.js

Commit:
added names to the exercise graph to show now instead of just their ID

saykaren commented:
Hello Brian @bpedersen! Patrick asked me to add you onto this pull request from my GetFitter repository. Thank you for the opportunity to show you my code and any feedback on how I can improve my skills.
Kindly,
Karen

bpedersen commented on server/getfitapi/app.js:
```diff
@@ -0,0 +1,44 @@
+
+var app = express();
+
+app.use(cors({origin: 'http://localhost:3001'}));
```
Might want to wrap this in a ENVIRONMENT_VARIABLE check so that this will work in production. This currently will only accept requests from localhost:3001

saykaren replied:
Thank you Brian that makes a lot of sense!

bpedersen commented on server/getfitapi/app.js:
```diff
@@ -0,0 +1,44 @@
+app.set('views', path.join(__dirname, 'views'));
+app.set('view engine', 'jade');
+
+app.use(logger('dev'));
```
Might want to use the ENVIRONMENT_VARIABLE here as well. But these are things you don't really need to address yet, but want to keep in mind for the future.

saykaren replied:
Thank you Brian I really appreciate your feedback as I didn't think about that in production mode.

bpedersen reviewed (changes requested):
Will you add server/getfitapi/node_modules to the .gitignore file so they are not included in the code. I'm happy to keep reviewing once those files are removed as that is the majority of the files included in this PR

bpedersen commented:
> Hello Brian @bpedersen! Patrick asked me to add you onto this pull request from my GetFitter repository. Thank you for the opportunity to show you my code and any feedback on how I can improve my skills.
> Kindly,
> Karen

No problem. I'm super happy to help as I started just like you. Please continue to send stuff my direction. Sometimes I may not see the github alerts, but I will try to stay on top of them.

Brian Pedersen

saykaren commented:
> Will you add server/getfitapi/node_modules to the .gitignore file so they are not included in the code. I'm happy to keep reviewing once those files are removed as that is the majority of the files included in this PR

@bpedersen -Yes! Sorry about that, it is now removed and added to the .gitignore files :)

saykaren commented:
> > Hello Brian @bpedersen! Patrick asked me to add you onto this pull request from my GetFitter repository. Thank you for the opportunity to show you my code and any feedback on how I can improve my skills.
> > Kindly,
> > Karen
> 
> No problem. I'm super happy to help as I started just like you. Please continue to send stuff my direction. Sometimes I may not see the github alerts, but I will try to stay on top of them.
> 
> Brian Pedersen

@bpedersen Thank you Brian! That is reassuring to know you started down the same path. I will absorb as much as you are willing to teach me so I can get better and continue to code as I am certainly enjoying this journey

bpedersen commented:
> > Will you add server/getfitapi/node_modules to the .gitignore file so they are not included in the code. I'm happy to keep reviewing once those files are removed as that is the majority of the files included in this PR
> 
> @bpedersen -Yes! Sorry about that, it is now removed and added to the .gitignore files :)

Ah, so this is a lesson in .gitignore. The tricky thing about adding the files now is that .gitignore will not touch them, but since they are already in the code they will not be removed. So to resolve this you need to remove the folder from the .gitignore, commit, delete the node_modules folder, add the folder back to the .gitignore, and then finally recommit.

@bpedersen Sorry Brian. I removed the node_modules from the GF-3/Feature but not the GF-2/Feature branch. Both should have that removed, let me know if I am wrong.

Commit:
removed the server/node_modules from GF-2/Feature branch

bpedersen commented on server/getfitapi/routes/users.js:
```diff
@@ -0,0 +1,696 @@
+/* GET users listing. */
+router.get('/', function(req, res, next) {
+  
+  let dataArray = [
```
You should split this array out to a .json file and require it in. It will help keep the temporary data seperate from your code.

saykaren replied:
This has been updates to have the user information in a .json file (user.json) and I required it into routes.js file that I made int he routes folder to be imported into the app.js  for the server as my usersRouter. Using my local server for my server and app seems to be working well. Pushed this update to the GF-2/Feature branch

bpedersen commented on src/Components/App/App.js:
```diff
@@ -1,51 +1,95 @@
+      res
+      .json()
+      .then(res => setUsers(res))
+      .then(res => setId(res[0].Id))
```
Does this actually set the id correctly? This then should have only the response value from setUsers, not the original res.

saykaren replied:
Hello Brian, testing it doesn't seem to set the ID so I removed and working well. I updated, committed and pushed to that branch for that update.

bpedersen commented on src/Components/App/App.js:
```diff
@@ -1,51 +1,95 @@
+  console.log({friends});
+
+  useEffect(() => {
+    async function fetchData() {
```
this function should be declared outside the scope of useEffect, otherwise you are recreating it each time that useEffect is called. It shouldn't matter in this case as the subscription for the useEffect is only on mount, but best practice is to declare this outside of the useEffect

saykaren replied:
@bpedersen Thank you for this insight. I did move fetchData function outside useEffect and seems to work well. I made the change and pushed it to the GF-2/Feature branch

bpedersen commented on src/Components/App/App.js:
```diff
@@ -1,51 +1,95 @@
+
+
   const userChange = event => {
     const stringEvent = event.target.value;
```
You should use event.currentTarget by default, unless you specifically have a reason for event.target (i.e. you have one handler, but you want to know which child was clicked) https://codepen.io/vikas-parashar/full/WZxrYX/ is a good codepen that shows the difference between currentTarget and target

saykaren replied:
Thank you @bpedersen Brian that is helpful to know and the codepen is a good visual. Updated and pushed to GF-2/Feature branch

bpedersen commented on src/Components/App/App.js:
```diff
@@ -1,51 +1,95 @@
   
-    if (eventTarget> 0 && eventTarget < userData.length-1){
-      var userResultArray = userData.filter(x=>(x.id === eventTarget));
+    if (eventTarget> 0 && eventTarget < users["data"].length){
```
Is the user.id never 0? Indexes normally start at 0

saykaren replied:
Hello Brian I am so happy to hear from you! @bpedersen That section has been updated a with a suggestion from Patrick. He introduced me to the find() method and doing object destructing to improve that function. 

Otherwise yes that was an error on my part as I was having the id of 0 be a guest object on the App.js file and then my app would search for other ids on the API.

Commit:
Per Brian's suggestion split the userData array into a .json file and required it into the now routes.js file I used for usersRouter

Commit:
On src/Components/App/App.js Brian noted that the .then(res => setID(res[0].Id) on line 47 doesn't actually set the ID and unnecessary. Removed and tested working great

Commit:
moved async function fetchData outside of useEffect as suggested by Brian as it recreating it each time useEffect is called. Subscription for the useEffect is only on the mount.

Commit:
changed App.js line 53 to event.currentTarget.value rather than event.target.value as suggested by Brian as a good default unless I have a reason for event.target directly (knowing which child was clicked)

patrickrauls commented on src/Components/App/App.js:
```diff
@@ -1,51 +1,86 @@
- 
   const userChange = event => {
-    const stringEvent = event.target.value;
+    const stringEvent = event.currentTarget.value;
```
naming things is one of the more difficult disciplines of this profession. It's important, as best you can, to give variables names that indicate what they are. To my knowledge, there is no native "string" event. it looks like this is actually collecting the value of a user id input any time that changes. As such, I would combine line 53 and 54 into `const userId = parseInt(value)`

saykaren replied:
@Patrickrauls Thank you Patrick! I agree I do need to get better at naming to be more descriptive as well as combining. Appreciate you looking at my code. I edited that and pushed that change onto the GF-2/Feature branch

patrickrauls commented on src/Components/App/App.js:
```diff
@@ -1,51 +1,86 @@
-  const [email, setEmail] = useState("");
-  const [friends, setFriends] = useState(userData[0].friends);
- 
   const userChange = event => {
```
Here I would consider doing something we call object destructuring. Using this we can essentially open a window into the argument given to the function and pull out exactly what we will use from that argument. We can do this as many levels deep as we want. It's particularly useful when we're given an object that has a bunch of stuff in it, but we only will use one or two tiny pieces of it. Here we would say `const userChange = ({ currentTarget: { value }}) => {`

By doing this we are actually declaring a variable called value and setting it equal event.target.value. So we're going two levels deep.

saykaren replied:
@Patrickrauls I think I understood this but let me know if I am wrong. So you are bringing in the value of the currentTarget as an object removing the need to declare it in the function. I just then did the userIdEvent line to change that object string into a number with parseInt()
const userChange = ({ currentTarget: { value }}) => {
    const userIdEvent = parseInt(value);. I updated the emailChange too as that was doing the same thing.

patrickrauls replied:
That's correct. The argument being passed to `userChange` is an `Event`. `Event`s are already objects, so we can look inside the `Event` object with [object destructuring](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#Object_destructuring) (the curly braces accomplish that). It turns out the value at `Event.currentTarget` is also an object, so we can look inside that object with more curly braces.

patrickrauls commented on src/Components/App/App.js:
```diff
@@ -1,51 +1,86 @@
-    if (eventTarget> 0 && eventTarget < userData.length-1){
-      var userResultArray = userData.filter(x=>(x.id === eventTarget));
+    if (eventTarget> 0 && eventTarget < users.length-1){
+      var userResultArray =  users.filter(x=>(x.id === eventTarget));
```
Line 57 - 61 we can do some interesting stuff. 1) 99% you'll probably want to use `const`. Only when you're creating a variable that you will reassign (set it equal to something else) later, should you use `let`. And you probably never need to use var in front end stuff. 2) We can actually replace line 57 - 61 by using a different Array method known as [find|https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find] and doing some destructuring. Instead of going through the whole array and returning a new array every element that passes a given test, find returns the first element to pass the test and bails out when it finds something. I assume you can assume `id`'s are unique across your data set, so you'll only match one. try 
`const { id, email, friends, name } = users.find(user => user.id === userId);`

patrickrauls replied:
of course you'll want to update all the arguments you're passing to your setState methods.

saykaren replied:
Thank you Patrick @patrickrauls! I did not know about the find() method that does work very well. Also yes the vars are a bad habit I learned that I need to get away from. Those changes have been made and pushed up to the GF-2/Feature branch. Thanks!

patrickrauls commented on src/Components/App/App.js:
```diff
@@ -1,51 +1,86 @@
+      const userResultName = userResultArray[0].name;
       setId(userResultId);
       setEmail(userResultEmail);
       changeContent(userResultId);
```
looks like you can remove this line

saykaren replied:
Thank you, it has been removed, especially with your suggested edit above with the find method I removed those sections as I no longer needed them.

patrickrauls commented on src/Components/App/App.js:
```diff
@@ -1,51 +1,86 @@
     };
   };
 
   const changeContent = (id)=>{
```
Not sure if you were using this method anywhere but on 64. If you weren't, you can get rid of it altogether. If you use VS Code it will give you some hinting when you have unused variables.

saykaren replied:
LOL yes after the previous method you showed me of find() I realized that I wasn't using this changeContent except to change the Id once again. It was duplication that I didn't need. Yes, VS code is very helpful with that, unfortunately I had it called in a function that was already updating the Id so this was over kill on my part. Thank you for catching that.

patrickrauls commented on src/Components/App/App.js:
```diff
@@ -1,51 +1,86 @@
   }
 
   //Ability for user to input email to change ID and name
   const emailChange = (event) => {
```
we're only using the value of this event's target, so another opportunity to destructure.

patrickrauls replied:
what is supposed to happen if you type something that doesn't match any email in user data?  Like for instance, while you're still in the middle of typing the email, so it doesn't technically match anything.

saykaren replied:
Oh good catch. I updated that code as well so my app wouldn't crash anymore it sets setId(0) setName("Wrong email"). Thanks for catching that. It is updated and pushed to GF-2/Feature @patrickrauls

patrickrauls commented on src/Components/App/App.js:
```diff
@@ -1,51 +1,86 @@
   const emailChange = (event) => {
     var stringEvent = event.target.value;
-    var userResultArray = userData.filter(x=>(x.email === stringEvent));
+    var userResultArray =  users.filter(x=>(x.email === stringEvent));
```
Array.prototype.find is your friend here.

saykaren replied:
LOL yes I updated this with the question about what if someone is typing in their email that doesn't match. I added the users.find method you told me about. Thank you!

patrickrauls commented on src/Components/App/App.js:
```diff
@@ -57,7 +92,7 @@ const App = () => {
       setEmail("");
```
should this be `setEmail(value)`? that's assuming that you've gone back and destructured the event.

saykaren replied:
Yes I updated that so you can see what you are typing and once it shows it is a valid email in the database it will then update the other states

patrickrauls commented on src/Components/App/FriendLookUp.js:
```diff
@@ -1,20 +1,38 @@
 import React from 'react';
-import userData from '../../data/users';
+// import userData from '../../data/users';
```
let's get rid of all of the commented out code.

saykaren replied:
Done :)

Commit:
Patrick suggested combining line 53 and 54 on App.js and renaming the variable to be more descriptive

Commit:
updated userChange and emailChange to do object destructuring with the currentTarget suggested by Patrick

Commit:
Used the find method Patrick introduced me to for filtering through an array and returning the first item that matches as I will only have one person associated with a unique ID. Also removed the vars, bad habit of mine

Commit:
Patrick realized I missed an area that if someone was typing in an email that didn't match it was crashing my app. Put if statements to check before it changes the state to ensure it matches something

Commit:
updated email change to have it set the Email value to what you enter in but it won't change anything else except for name until it matches something in the database

patrickrauls commented on src/Components/App/App.js:
```diff
@@ -1,63 +1,85 @@
-      setName("Wrong email");
-      setEmail("");
+  const userChange = ({ currentTarget: { value }}) => {
+    const userIdEvent = parseInt(value);
```
What is your thinking in including "Event" in this variable name?

saykaren replied:
@patrickrauls I just have so many variables with user in it I wanted to differentiate it. I have userId in the data, I have users as state and I have Id as state that I just wanted to differentiate it a bit more so I didn't confuse it with the state or userId in the data objects. I wanted to clarify for myself that it was the userId I received from this event that is happening which is someone changing the ID on app. Was the wrong to add event to the end?

patrickrauls replied:
To me it's somewhat misleading in that it sort of sounds like there is some userId event being kicked off in the browser. I'd be interested to get the opinion of @bpedersen on that in case I'm off base. 

At any rate, I think it's fine to leave it for now. I suspect we'll eventually make some changes that will address the underlying reason you had to have so many variables holding similar info.

patrickrauls commented on src/Components/App/App.js:
```diff
@@ -1,63 +1,85 @@
 
+  //Ability for user to input email to change ID and name
+  const emailChange = ({ currentTarget: { value }}) => {
+    const validEmail = users.find(user => user.email === value); 
```
consider naming this variable something else. users.find isn't returning an email, it's returning a user. I would use something like `user` or `foundUser`. Something to that effect.

saykaren replied:
@patrickrauls Thank you Patrick! That is helpful to think more on what I name my variables as yes this is returning an object of the user not just the email itself. I updated that and pushed it to this branch.

patrickrauls commented on src/Components/App/App.js:
```diff
@@ -1,63 +1,85 @@
+  const emailChange = ({ currentTarget: { value }}) => {
+    const validEmail = users.find(user => user.email === value); 
+    if (typeof(validEmail) !== "undefined" && typeof(validEmail) !==null){
+      const {id, email, name } = users.find(user => user.email === value);
```
you've already done the work of finding the user in line 69. Because you've stored that result in the variable above, you can use [object destructuring](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#Object_destructuring) to pull it from the variable here. 
`const { id, email, name } = foundUser;`

saykaren replied:
@patrickrauls Perfect! Thank you. I had both as if I put the const {id, email, name} =... before the if statement it was crashing my app as with TypeError: Cannot read property 'id' of undefined". I didn't think to do the object destructuring after the if statement from foundUser rather than doing another find method. Thank you for your help!
Added this change and pushed it to the GF-2/Feature branch

Commit:
updated variable name on line 69 of App.js to be more descriptive of foundUser as it isn't just the email it is the object of a user that was found that matched that criteria

Commit:
In App.js Patrick noted that I don't need to do the find method again in emailChange function I just need to do object destructuring of foundUser after the if statement. I already did the work to find that match I shouldn't have it do it again just destructure it after I verify with if statement it is a valid object

patrickrauls commented on src/Components/App/App.js:
```diff
@@ -1,63 +1,85 @@
+  const userChange = ({ currentTarget: { value }}) => {
+    const userIdEvent = parseInt(value);
+
+    if (userIdEvent> 0 && userIdEvent < users.length-1){
```
This works. But I think there is a way to get it a bit more clear. Really what you want to know is, "Does the users array contain a user with the id that we just pulled from the input?". [Array.prototype.some](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some) can tell you exactly that.

saykaren replied:
```
  const userChange = ({ currentTarget: { value }}) => {
    const userId = parseInt(value);
    const testValidId = (e) =>{return e.id == userId;}
   
    const userIdValid =(e)=>{
      const { id, email, friends, name } = users.find(user => user.id === e);
      setId(id);
      setEmail(email);
      setFriends(friends);
      setName(name);  
    };

    const userIdInvalid = () =>{
      setId(guestData[0].id);
      setName(guestData[0].name);
      setEmail(guestData[0].email);
    };
 
    (users.some(testValidId)) ? userIdValid(userId) : userIdInvalid();
  }

```

@patrickrauls Thank you for the help. I hope I am not going off the deep end here. Since Array.prototype.some() requires a function I made the testValidID function so that I could do the users.some with testValidID as the argument. Then I thought I could use the new syntax at the end to valuate that and if it is true do the function userIdValid passing through the userId and if not doing the userIdInvalid function that mostly just sets all the data back to a baseline of guest. Let me know if I am off base with this and I can switch it back. I am going to keep my old code commented out until you let me know if I am on the right path :)

patrickrauls commented on src/Components/App/App.js:
```diff
@@ -1,63 +1,85 @@
 
 const App = () => {
+  
+  const guestData = [
```
unless you intend to have several guests at some point, you don't have to hold this in an array.

saykaren replied:
@patrickrauls Thank you Patrick. I had it as a template to start my states with but I removed it and updated my states to reflect that.

Commit:
line 54 Patrick suggested to do Array.prototype.some() function. I changed a lot of that userChange function code based upon that. Asking Patrick if I am on the right track or if I went down the wrong rabbit hole.

Commit:
Per Patrick suggested removed guestData in App.js itself. Had to adjust some of the initial states as they were based upon that to start up the app but now it is imports that data and then bases it upon that information.

Commit:
Thought I removed it before but obviously didn't commit it right for removing users.js data, removed scripts file as that was one I started with using classes but went on from that. Deleted the server.js file as I did that under the server folder with the express API.
